doc/en/stilview.1
'\" t
.\"     Title: stilview
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.77.1 <http://docbook.sf.net/>
.\"      Date: September 2012
.\"    Manual: Reference
.\"    Source: stilview
.\"  Language: English
.\"
.TH "STILVIEW" "1" "September 2012" "stilview" "Reference"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
stilview \- command\-line program to help you retrieve the entries stored in STIL
.SH "SYNOPSIS"
.HP \w'\fBstilview\fR\ 'u
\fBstilview\fR [\fB\-b\fR] [\fB\-d\fR] [\fB\-e\fR\ \fIentry\fR] [\fB\-f\fR\ \fIfield\fR] [\fB\-i\fR] [\fB\-l\fR\ \fIHVSC\ base\ dir\fR] [\fB\-m\fR] [\fB\-o\fR] [\fB\-s\fR] [\fB\-t\fR\ \fItune\ number\fR]
.HP \w'\fBstilview\fR\ 'u
\fBstilview\fR {[\-h] | [\-v]}
.SH "DESCRIPTION"
.PP
STILView is a command\-line driven program to help you retrieve the entries stored in STIL fast and accurately\&. STILView uses the STIL C++ class heavily to do this, and in fact, the primary purpose of this command\-line program is to test that class (which is, BTW, used in many GUI\-based SID players, most notably in SIDPlay for Windows and XSIDPLAY for Unix systems)\&. However, it is user\-friendly enough to be used by non\-programmers, too\&.
.PP
I am not sure about what other people do, but I use this command\-line STILView primarily as a tool under Solaris 2\&.6 to print out the STIL information for a SID file I am listening to with the UNIX command\-line version of SIDPlay\&.\&.\&.
.SH "GLOSSARY"
.PP
Some terms and STIL\-related lingo in alphabetical order:
.PP
BUG ENTRY \- There exists a special file in HVSC (/DOCUMENTS/BUGlist\&.txt) that lists all known bugs in SID tunes in HVSC\&. See the top of that file for details about what\*(Aqs in it exactly\&. A BUG entry is like a STIL entry, but it is contained in this BUGlist\&.txt file\&.
.PP
FIELD \- The smallest piece of information in a STIL entry\&. Currently valid field names are NAME, TITLE, ARTIST and COMMENT\&.
.PP
FILE\-GLOBAL COMMENT \- A special COMMENT field in a STIL entry for a multi\-tune SID file that refers to the whole SID, not just one tune in it\&. These usually contain general information about the SID file itself\&.
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
/Hubbard_Rob/Gerry_the_Germ\&.sid
COMMENT: In Rob\*(Aqs own demo of this music, the tunes are named after the levels
         in the original game\&.
(#1)
  TITLE: Lungs
(#2)
  TITLE: Kidney
(#7)
  TITLE: End
.fi
.if n \{\
.RE
.\}
.PP
HVSC \- High Voltage SID Collection\&. If you don\*(Aqt know what this is, you downloaded the wrong program\&. :)
.PP
HVSC\-RELATIVE PATHNAME \- The pathname plus filename of a SID file that can be found in your HVSC, relative to the base directory of HVSC\&. It is always in UNIX\-style format, eg\&.: /Hubbard_Rob/Commando\&.sid refers to Rob Hubbard\*(Aqs Commando\&.sid file within HVSC (which may actually be found as C:\eMusic\eHVSC\eHubbard_Rob\eCommando\&.sid on your Windows PC)\&.
.PP
MULTI\-TUNE ENTRY \- A STIL entry that is referring to a SID file that has many tunes in it\&. Each tune might have its own STIL block, which are separated by a so\-called tune designation in the form of "(#x)", where x = the tune number\&. Consult the STIL\&.FAQ in HVSC for a detailed description\&.
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
/Hubbard_Rob/Gerry_the_Germ\&.sid
COMMENT: In Rob\*(Aqs own demo of this music, the tunes are named after the levels
         in the original game\&.
(#1)
  TITLE: Lungs
(#2)
  TITLE: Kidney
(#7)
  TITLE: End
.fi
.if n \{\
.RE
.\}
.PP
SECTION \- A part of STIL that belongs to one composer (ie\&. every STIL entry referring to SID files that are in one subdirectory in HVSC)\&. Sections in STIL are always separated by a line in the form of: "### Composer\*(Aqs name ########"\&.
.PP
SECTION\-GLOBAL COMMENT \- A special STIL entry that refers not to an individual SID file, but to a whole subdirectory\&. These usually contain info about the composer himself, or about all the SID file he/she ever composed, and are always indexed in the form of "/Subdir/" (note the trailing slash!)\&.
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
/Hubbard_Rob/
COMMENT: All of these tunes have been confirmed by Hubbard to be his\&. People
         have often stolen Hubbard\*(Aqs routine causing some tunes to be falsely
         credited to him\&.
         Hubbard\*(Aqs own comments are denoted by (RH)\&.
.fi
.if n \{\
.RE
.\}
.PP
SINGLE\-TUNE ENTRY \- A STIL entry that has no tune designation in it in the form of "(#x)", where x is a number\&. (Note, that a single\-tune entry might still refer to a SID file which has many tunes in it, ie\&. when a single\-tune entry has nothing but a COMMENT field in it!)
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
/Hubbard_Rob/Chain_Reaction\&.sid
  TITLE: Zoolook (remix) [from Zoolook]
 ARTIST: Jean Michel Jarre
.fi
.if n \{\
.RE
.\}
.PP
Another example (the SID this is refering to has many tunes in it!):
.sp
.if n \{\
.RS 4
.\}
.nf
/Barrett_Steve/Magic_Land_Dizzy\&.sid
COMMENT: Also used in the game "Wacky Darts" (c) 1990 Codemasters\&.
.fi
.if n \{\
.RE
.\}
.PP
STIL \- SID Tune Information List, essentially a text\-file database that can be found in your HVSC in the /DOCUMENTS/ subdirectory\&.
.PP
STIL ENTRY \- All of the pieces of information in STIL relating to one SID file of the HVSC\&. They are always indexed by the HVSC\-relative pathname\&.
.PP
TUNE \- One of the compositions in a SID\&. Most SID files have only one tune in them, but many have more than one (eg\&. one for the title score of the game, and one for the hi\-score music)\&.
.SH "OPTIONS"
.PP
\fB\-b\fR
.RS 4
Do *not* print BUG entries
.sp
Default value: Not specified (ie\&. do print BUG entries) Example: stilview \-e=/Hubbard_Rob/Commando\&.sid \-b
.sp
When this option is specified, BUG entries will not be printed for the given SID tune\&. At a minimum, the \-e option has to be specified for this option to work\&.
.RE
.PP
\fB\-d\fR
.RS 4
Default value: Not specified (ie\&. debug mode is off)
.sp
Example: stilview \-e=/Hubbard_Rob/Commando\&.sid \-d
.sp
Turns on debug mode in STILView\&. This will result in an extensive output, with the debugging messages going to STDERR\&. If you encounter any problem or strange behavior with STILView, run STILView with the exact same options as you did when you encountered the problem, with this \-d option added to them\&. Capture the complete output of this run, and send it to me with a detailed explanation of the problem (see email address at the top of this file)\&.
.RE
.PP
\fB\-e\fR=\fIentry\fR
.RS 4
Default: NONE (you have to give an HVSC\-relative pathname to this option)
.sp
Example #1: stilview \-e=/Hubbard_Rob/Commando\&.sid Example #2: stilview \-e=/Hubbard_Rob/
.sp
This is where you specify the STIL entry you are looking for, given as an HVSC\-relative pathname\&. If there exists no STIL entry for the given filename, STILView will print out nothing\&. Otherwise, you\*(Aqll get the STIL entry (or parts of it, as you may have specified it by other options)\&. HVSC\-relative pathnames are case\-insensitive, so /HUBBARD_ROB/Commando\&.sid is the same as /Hubbard_Rob/Commando\&.sid\&.
.sp
Example #1 is the most frequent way of retrieving STIL entries, and it will return all of the STIL entry for Commando\&.sid, as well as the section\-global comment for /Hubbard_Rob/\&. Example #2 is another valid thing to do: this will return only the section\-global comment for /Hubbard_Rob/\&.
.RE
.PP
\fB\-h\fR
.RS 4
Default: NONE Example: stilview \-h
.sp
Prints a brief help screen listing the available options\&. All other options that are also specified on the command\-line are ignored\&.
.RE
.PP
\fB\-f\fR=\fIfield\fR
.RS 4
Default: all
.sp
Valid values for <field> are: all, name, author, title, artist, comment
.sp
Example #1: stilview \-l \-e=/Hubbard_Rob/Delta\&.sid \-f=comment
.sp
Example #2: stilview \-l \-e=/Hubbard_Rob/Delta\&.sid \-t=1 \-f=title
.sp
Example #3: stilview \-l \-e=/Hubbard_Rob/Delta\&.sid \-t=12 \-f=all \-s \-b
.sp
Asks for one particular field in a STIL entry\&. Combined with the \-t option, these two options can retrieve any portion of a STIL entry, including a single field in a specific subtune\*(Aqs entry\&. Below is full and complete explanation of what the different possible combinations of the \-t and \-f options retrieve:
.sp
\-t=0 \-f=all : All of the STIL entry is printed\&.
.sp
\-t=0 \-f=comment : The file\-global comment is printed\&. For single\-tune entries that have nothing but a COMMENT field in them, this prints that COMMENT\&. For single\-tune entries that have other fields in them, this prints nothing\&. (This is because single\-tune entries with nothing but a COMMENT field are assumed to be file\-global comments\&.)
.sp
\-t=0 \-f=<name/author/title/artist> : Nothing is printed\&. This combination of these options is invalid\&.
.sp
\-t=<x> \-f=all : (Where x is anything but 0\&.) All fields from the portion of the STIL entry for the given tune number <x> are printed\&. For single\-tune entries, asking for \-t=1 \-f=all is equivalent to saying \-t=0 \-f=all, since by definition, the whole entry refers to only one tune\&. (However, specifying \-t with any other number than 1 will print nothing!) Note that if there\*(Aqs a file\-global comment in the STIL entry (which also means that if a single\-tune entry has nothing but a COMMENT field in it), that is *not* printed with these combinations of options\&.
.sp
\-t=<x> \-f=<name/author/title/artist/comment> : (Where x is anything but 0\&.) The specific field from the portion of the STIL entry for the given tune number is printed\&. For single\-tune entries that have nothing but a COMMENT in them, this returns nothing\&.
.sp
Of course, if the STIL entry or any portion of it asked with these options does not exist, STILView will print nothing\&. Also, unless otherwise specified with the \-o, \-s and \-b options, the section\-global comment and the BUG entry of the given SID file will also get printed (provided they exist)\&.
.sp
In example #1, the file\-global comment for /Hubbard_Rob/Delta\&.sid is printed, since \-t is not specified and is assumed to be 0\&. Also printed are the section\- global comment and the BUG entry for the same SID file (if they exist)\&. In example #2, the TITLE field of the STIL entry for tune #1 of /Hubbard_Rob/Delta\&.sid is printed along with the section\-global comment and the BUG entry for the same SID file (if they exist)\&. In example #3, all of the STIL entry for tune #12 of /Hubbard_Rob/Delta\&.sid is printed, but nothing else\&.
.RE
.PP
\fB\-i\fR
.RS 4
Default: NONE
.sp
Example: stilview \-i
.sp
Starts STILView in interactive mode, ignoring all other options specified on the command\-line, except \-l, \-d and \-m\&. In interactive mode, you can look for STIL entries by typing them in\&. You will get prompted for the desired STIL entry (which has to be specified with an HVSC\-relative pathname), for the tune number requested (which should be any non\-negative number, but this is not enforced), and finally for the specific STIL field you want to retrieve\&.
.RE
.PP
\fB\-l\fR=\fIHVSC base dir\fR
.RS 4
Default: The value of the HVSC_BASE environment variable
.sp
Example #1: stilview \-l=C:\eMusic\eHVSC\e \-e=/Hubbard_Rob/Commando\&.sid
.sp
Example #2: stilview \-l=\&.\&./HVSC/ =\-e=/Hubbard_Rob/Commando\&.sid
.sp
Example #3: stilview \-l \-e=/Hubbard_Rob/Commando\&.sid
.sp
This is where you tell STILView where it can find the HVSC base directory (the path to the directory has to be specified in the form required by your operating system, eg\&. C:\eMusic\eHVSC under Windows, /home/lala/HVSC under UNIX)\&. STILView will then try to locate the STIL\&.txt file in the /DOCUMENTS/ subdirectory of that directory\&. If this option is not specified (or if \-l is specified without a base directory), STILView will try to extract the path of the HVSC base directory from the HVSC_BASE environment variable\&. If that environment variable doesn\*(Aqt exist or is pointing to a location where there\*(Aqs no STIL\&.txt file in a DOCUMENTS directory, STILView fails\&. If the HVSC_BASE environment variable exists and is valid, and this option is specified, the directory specified with this option is used as the HVSC base directory instead of the environment variable\&.
.sp
In example #1 the HVSC base directory is located in C:\eMusic\eHVSC\e on the hard drive of a Windows PC, in example #2 it is located in the HVSC directory of the current directory\*(Aqs parent directory of a UNIX system\&. In example #3 the HVSC base directory is not specified with the option, so it is assumed that the HVSC_BASE environment variable contains the path to it\&. In reality, specifying the \-l option in example #3 is redundant, and can be omitted\&.
.RE
.PP
\fB\-m\fR
.RS 4
Demo mode
.sp
Default: NONE
.sp
Example #1: stilview \-m
.sp
Example #2: stilview \-e=/Hubbard_Rob/Commando\&.sid \-m \-i
.sp
When specified, it prints out a whole bunch of things that a) test most of the functionality of STILView, and b) show what STILView is capable of retrieving from STIL\&. In example #1, the demo is printed with the STIL info coming from a default STIL entry, then STILView quits\&. In example #2, the demo is printed taking the STIL info from the specified STIL entry of /Hubbard_Rob/Commando\&.sid (instead of the default SID file), then interactive mode is entered\&.
.RE
.PP
\fB\-o\fR
.RS 4
Do *not* print STIL entries
.sp
Default value: Not specified (ie\&. do print STIL entries)
.sp
Example #1: stilview \-e=/Hubbard_Rob/Delta\&.sid \-o
.sp
Example #2: stilview \-e=/Hubbard_Rob/Delta\&.sid \-o \-s
.sp
When this option is specified, STIL entries will not be printed for the given SID tune (but section\-global entries and BUG entries will be printed, provided they exist and other options did not turn their output off)\&. At a minimum, the \-e option has to be specified for this option to work\&. Example #1 will print out the section\-global comment and the BUG entry for /Hubbard_Rob/Delta\&.sid, example #2 will print out just the section\-global comment for the same SID\&.
.RE
.PP
\fB\-s\fR
.RS 4
Do *not* print section\-global comments
.sp
Default value: Not specified (ie\&. do print section\-global entries)
.sp
Example: stilview \-e=/Hubbard_Rob/Delta\&.sid \-s
.sp
When this option is specified, section\-global entries will not be printed for the given SID tune\&. At a minimum, the \-e option has to be specified for this option to work\&.
.RE
.PP
\fB\-t\fR=\fItune number\fR
.RS 4
Default value: 0
.sp
Example #1: stilview \-e=/Hubbard_Rob/Commando\&.sid \-t=0
.sp
Example #2: stilview \-e=/Hubbard_Rob/Delta\&.sid \-t=1 \-f=title \-s \-b
.sp

src/stilview.cpp
/*
 * This file is part of sidplayfp, a console SID player.
 *
 * Copyright 2012 Leandro Nini
 * Copyright 1998, 2002 LaLa <LaLa@C64.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

//
// STILView - command line version
//

#include <fstream>
#include <iostream>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "stilview/stil.h"

STIL myStil;
char *hvscLoc = NULL;
char *entryStr = NULL;
int tuneNo = 0;
STIL::STILField field = STIL::all;
bool showBug = true;
bool showEntry = true;
bool showSection = true;
bool showVersion = false;
bool interactive = false;
bool demo = false;

char STIL_DEMO_ENTRY[]="/Galway_Martin/Green_Beret.sid";

// This is used for testing setBaseDir() when switching between different
// HVSC base directories. Ideally, it should point to a valid HVSC dir.
char OTHER_HVSC_BASE_DIR[]="E:\\MUSIC\\SID\\C64music\\";

#define STIL_MAX_PATH_SIZE 1024

using namespace std;

char toLowerAscii(char c)
{
    return (c < 'A' || c > 'Z') ? c : c + ('a' - 'A');
}

bool strEquals(const char* s1, const char* s2)
{
    while (*s1 && *s2)
    {
        if (toLowerAscii(*s1) != toLowerAscii(*s2))
            return false;
        s1++;
        s2++;
    }

    return !(*s1 || *s2);
}

void printUsageStr(void)
{
    cout << endl;
    cout << myStil.getVersion();
    cout << "USAGE: STILView [-e=<entry>] [-l=<HVSC loc>] [-t=<tuneNo>] [-f=<field>]" << endl;
    cout << "                [-d] [-i] [-s] [-b] [-o] [-v] [-h] [-m]" << endl;
}

void printUsage(void)
{
    printUsageStr();
    exit(1);
}

void printHelp(void)
{
    printUsageStr();
    cout << "Arguments can be specified in any order." << endl;
    cout << endl;
    cout << "-e=<entry>    - Specifies the desired STIL entry with an HVSC-relative path." << endl;
    cout << "-l=<HVSC loc> - Specifies the location of the HVSC base directory. If not" << endl;
    cout << "                specified, the value of the HVSC_BASE env. variable will be" << endl;
    cout << "                used. Specifying this option will override HVSC_BASE." << endl;
    cout << "-t=<tuneNo>   - If specified, only the STIL entry for the given tune number is" << endl;
    cout << "                printed. (Default: 0)" << endl;
    cout << "-f=<field>    - If specified, only the STIL entry for the given field is" << endl;
    cout << "                printed. (Default: all)" << endl;
    cout << "                Valid values for <field> are:" << endl;
    cout << "                all, name, author, title, artist, comment" << endl;
    cout << "-d            - Turns on debug mode for STILView." << endl;
    cout << "-i            - Enter interactive mode." << endl;
    cout << "-m            - Demo mode (tests STILView and shows its capabilities)." << endl;
    cout << "-s            - If specified, section-global (per dir/per composer) comments" << endl;
    cout << "                will NOT be printed." << endl;
    cout << "-b            - If specified, BUG entries will NOT be printed." << endl;
    cout << "-o            - If specified, STIL entries will NOT be printed." << endl;
    cout << "-v            - Print STILView's and STIL's version number." << endl;
    cout << "-h            - Print this help screen (all other options are ignored)." << endl;
    cout << endl;
    cout << "See user manual for further details and for examples." << endl;
    cout << endl;
    exit(0);
}

char *getArgValue(char *argStr)
{
    char *temp;

    temp = (char *)strchr(argStr, '=');

    if (temp == NULL) {
        return NULL;
    }

    if (*(temp+1) == '\0') {
        return NULL;
    }

    return (temp+1);
}

void processArguments(int argc, char **argv)
{
    int i;
    char *fieldStr;
    char *tuneStr;
    char *tempLoc;

    for (i=1; i<argc; i++) {
        if (argv[i][0] == '-') {
            switch (argv[i][1]) {
                case 'd':
                case 'D':
                    myStil.STIL_DEBUG = true;
                    break;
                case 'e':
                case 'E':
                    entryStr = getArgValue(argv[i]);
                    if (entryStr == NULL) {
                        cerr << "ERROR: STIL entry was not specified correctly!" << endl;
                        printUsage();
                    }
                    break;
                case 'i':
                case 'I':
                    interactive = true;
                    break;
                case 'm':
                case 'M':
                    demo = true;
                    break;
                case 'l':
                case 'L':
                    tempLoc = getArgValue(argv[i]);
                    if (tempLoc != NULL) {
                        hvscLoc = tempLoc;
                    }
                    break;
                case 't':
                case 'T':
                    tuneStr = getArgValue(argv[i]);
                    if (tuneStr == NULL) {
                        cerr << "ERROR: tune number was not specified correctly!" << endl;
                        printUsage();
                    }
                    sscanf(tuneStr, "%d", &tuneNo);
                    break;
                case 's':
                case 'S':
                    showSection = false;
                    break;
                case 'b':
                case 'B':
                    showBug = false;
                    break;
                case 'o':
                case 'O':
                    showEntry = false;
                    break;
                case 'v':
                case 'V':
                    showVersion = true;
                    break;
                case 'f':
                case 'F':
                    fieldStr = getArgValue(argv[i]);
                    if (fieldStr == NULL) {
                        cerr << "ERROR: field was not specified correctly!" << endl;
                        printUsage();
                    }
                    if (strEquals(fieldStr, "all")) {
                        field = STIL::all;
                    }
                    else if (strEquals(fieldStr, "name")) {
                        field = STIL::name;
                    }
                    else if (strEquals(fieldStr, "author")) {
                        field = STIL::author;
                    }
                    else if (strEquals(fieldStr, "title")) {
                        field = STIL::title;
                    }
                    else if (strEquals(fieldStr, "artist")) {
                        field = STIL::artist;
                    }
                    else if (strEquals(fieldStr, "comment")) {
                        field = STIL::comment;
                    }
                    else {
                        cerr << "ERROR: Unknown STIL field specified: '" << fieldStr << "' !" << endl;
                        cerr << "Valid values for <field> are:" << endl;
                        cerr << "all, name, author, title, artist, comment." << endl;
                        printUsage();
                    }
                    break;
                case 'h':
                case 'H':
                    printHelp();
                    break;
                default:
                    cerr << "ERROR: Unknown argument: '" << argv[i] << "' !" << endl;
                    printUsage();
                    break;
            }
        }
        else {
            cerr << "ERROR: Unknown argument: '" << argv[i] << "' !" << endl;
            printUsage();
        }
    }
}

void checkArguments(void)
{
    if (hvscLoc == NULL) {
        if (interactive || demo) {
            hvscLoc = new char[STIL_MAX_PATH_SIZE];
            cout << "Enter HVSC base directory: ";
            cin >> hvscLoc;
        }
        else {
            if (showVersion) {
                showBug = false;
                showEntry = false;
                showSection = false;
            }
            else {
                cerr << "ERROR: HVSC base dir was not specified and HVSC_BASE is not set, either!" << endl;
                printUsage();
            }
        }
    }

    if (entryStr == NULL) {
        if ((!interactive) && (!demo)) {
            if (showVersion) {
                showBug = false;
                showEntry = false;
                showSection = false;
            }
            else {
                cerr << "ERROR: STIL entry was not specified!" << endl;
                printUsage();
            }
        }
        else {
            entryStr = STIL_DEMO_ENTRY;
        }
    }
}

int main(int argc, char **argv)
{
    char temp[STIL_MAX_PATH_SIZE];
    const char *tmpptr, *sectionPtr, *entryPtr, *bugPtr;
    const char *versionPtr;
    char fieldchar;
    float tempval;

    if (argc < 2) {
        printHelp();
    }

    hvscLoc = getenv("HVSC_BASE");

    processArguments(argc, argv);

    checkArguments();

    if (interactive || demo) {
        cout << "Reading STIL..." << endl;
    }
    else {
        if (showVersion && (hvscLoc == NULL)) {
            versionPtr = myStil.getVersion();
            if (versionPtr == NULL) {
                cerr << "ERROR: No STIL version string was found!" << endl;
            }
            else {
                cout << versionPtr;
            }

            exit(0);
        }
    }

    if (myStil.setBaseDir(hvscLoc) != true) {
        cerr << "STIL error #" << myStil.getError() << ": " << myStil.getErrorStr() << endl;
        exit(1);
    }

    if ((!interactive) && (!demo)) {

        // Pure command-line version.

        if (showVersion) {
            versionPtr = myStil.getVersion();
        }
        else {
            versionPtr = NULL;
        }

        if (showSection) {
            sectionPtr = myStil.getGlobalComment(entryStr);
        }
        else {
            sectionPtr = NULL;
        }

        if (showEntry) {
            entryPtr = myStil.getEntry(entryStr, tuneNo, field);
        }
        else {
            entryPtr = NULL;
        }

        if (showBug) {
            bugPtr = myStil.getBug(entryStr, tuneNo);
        }
        else {
            bugPtr = NULL;
        }

        if (versionPtr != NULL) {
            if ((sectionPtr != NULL) || (entryPtr != NULL) || (bugPtr != NULL)) {
                cout << "--- STILView  VERSION ---" << endl;
            }
            cout << versionPtr;
        }

        if (sectionPtr != NULL) {
            if ((versionPtr != NULL) || (entryPtr != NULL) || (bugPtr != NULL)) {
                cout << "---- GLOBAL  COMMENT ----" << endl;
            }
            cout << sectionPtr;
        }

        if (entryPtr != NULL) {
            if ((versionPtr != NULL) || (sectionPtr != NULL) || (bugPtr != NULL)) {
                cout << "------ STIL  ENTRY ------" << endl;
            }
            cout << entryPtr;
        }

        if (bugPtr != NULL) {
            if ((versionPtr != NULL) || (sectionPtr != NULL) || (entryPtr != NULL)) {
                cout << "---------- BUG ----------" << endl;
            }
            cout << bugPtr;
        }
    }
    else {

        // We are either in interactive or demo mode here.

        if (demo) {
            cout << "==== STILVIEW  DEMO MODE ====" << endl;
            cout << endl << "---- STIL VERSION ----" << endl;
            cout << "---- ONE STRING ----" << endl;
        }

        // This gets printed regardless.

        versionPtr = myStil.getVersion();
        if (versionPtr == NULL) {
            cerr << "ERROR: No STIL version string was found!" << endl;
        }
        else {
            cout << versionPtr;
        }

        if (demo) {

            // Demo mode.

            cout << "---- STIL CLASS VERSION # ----" << endl;
            tempval = myStil.getVersionNo();
            if (tempval == 0) {
                cerr << "ERROR: STILView version number was not found!" << endl;
            }
            else {
                cout << "STILView v" << tempval << endl;
            }

            cout << "---- STIL.txt VERSION # ----" << endl;
            tempval = myStil.getSTILVersionNo();
            if (tempval == 0) {
                cerr << "ERROR: STIL version number was not found!" << endl;
            }
            else {
                cout << "STIL v" << tempval << endl;
            }

            // For testing setBaseDir().

            if (myStil.STIL_DEBUG == true) {
                if (myStil.setBaseDir(OTHER_HVSC_BASE_DIR) != true) {
                    cerr << "STIL error #" << myStil.getError() << ": " << myStil.getErrorStr() << endl;
                    cerr << "Couldn't switch to new dir: '" << OTHER_HVSC_BASE_DIR << "'" << endl;
                    cerr << "Reverting back to '" << hvscLoc << "'" << endl;
                }
                else {
                    hvscLoc = OTHER_HVSC_BASE_DIR;

                    cout << "Switch to new dir '" << hvscLoc << "' was successful!" << endl;

                    cout << "---- ONE STRING ----" << endl;

                    versionPtr = myStil.getVersion();
                    if (versionPtr == NULL) {
                        cerr << "ERROR: No STIL version string was found!" << endl;
                    }
                    else {
                        cout << versionPtr;
                    }

                    cout << "---- STIL CLASS VERSION # ----" << endl;
                    tempval = myStil.getVersionNo();
                    if (tempval == 0) {
                        cerr << "ERROR: STILView version number was not found!" << endl;
                    }
                    else {
                        cout << "STILView v" << tempval << endl;
                    }

                    cout << "---- STIL.txt VERSION # ----" << endl;
                    tempval = myStil.getSTILVersionNo();
                    if (tempval == 0) {
                        cerr << "ERROR: STIL version number was not found!" << endl;
                    }
                    else {
                        cout << "STIL v" << tempval << endl;
                    }
                }
            }

            cout << endl << "==== STIL ABSOLUTE PATH TO " << entryStr << ", Tune #" << tuneNo << " ====" << endl << endl;

            strcpy(temp, hvscLoc);

            // Chop the trailing slash
            char *tmp = temp+strlen(temp)-1;
            if (*tmp == SLASH) {
                *tmp = '\0';
            }
            strcat(temp, entryStr);

            cout << "---- GLOBAL  COMMENT ----" << endl;

            tmpptr = myStil.getAbsGlobalComment(temp);

            if (tmpptr == NULL) {
                cerr << "STIL error #" << myStil.getError() << ": " << myStil.getErrorStr() << endl;
            }
            else {
                cout << tmpptr;
            }

            cout << "-- TUNE GLOBAL COMMENT --" << endl;

            tmpptr = myStil.getAbsEntry(temp, 0, STIL::comment);

            if (tmpptr == NULL) {
                cerr << "STIL error #" << myStil.getError() << ": " << myStil.getErrorStr() << endl;
            }
            else {
                cout << tmpptr;
            }

            cout << "------ STIL  ENTRY ------" << endl;
            cout << "(For tune #1)" << endl;

            tmpptr = myStil.getAbsEntry(temp, 1, STIL::all);

            if (tmpptr == NULL) {
                cerr << "STIL error #" << myStil.getError() << ": " << myStil.getErrorStr() << endl;
            }
            else {
                cout << tmpptr;
            }

            cout << "---------- BUG ----------" << endl;

            tmpptr = myStil.getAbsBug(temp, tuneNo);

            if (tmpptr == NULL) {
                cerr << "STIL error #" << myStil.getError() << ": " << myStil.getErrorStr() << endl;
            }
            else {
                cout << tmpptr;
            }

            cout << "==== END OF ENTRY ====" << endl;

            cout << endl << "Trying to do setBaseDir() to wrong location..." << endl;

            if (myStil.setBaseDir("This_should_not_work") != true) {
                cout << "setBaseDir() failed!" << endl;
                cout << "But it should't have an impact on private data!" << endl;
                cout << "You should see the same entry below:" << endl;
                cout << endl << "------ STIL  ENTRY ------" << endl;

                tmpptr = myStil.getAbsEntry(temp, tuneNo, STIL::all);

                if (tmpptr == NULL) {
                    cerr << "STIL error #" << myStil.getError() << ": " << myStil.getErrorStr() << endl;
                }
                else {
                    cout << tmpptr;
                }
            }
            else {
                cout << "Oops, it should've failed!" << endl;
            }
        }

        if (interactive) {

            // Interactive mode.

            cout << endl << "==== ENTERING INTERACTIVE MODE ====" << endl << endl;

            do {
                cout << "Enter desired entry (relative path) or 'q' to exit." << endl;
                cout << "Entry: ";
                cin >> temp;

                if (*temp == '/') {
                    cout << "Enter tune number (can enter 0, too): ";
                    cin >> tuneNo;

                    cout << "Field [(A)ll, (N)ame, A(U)thor (T)itle, A(R)tist,(C)omment]: ";
                    cin >> fieldchar;

                    switch (fieldchar) {
                        case 'a':
                        case 'A':
                            field = STIL::all;
                            break;
                        case 'n':
                        case 'N':
                            field = STIL::name;
                            break;
                        case 'u':
                        case 'U':
                            field = STIL::author;
                            break;
                        case 't':
                        case 'T':
                            field = STIL::title;
                            break;
                        case 'r':
                        case 'R':
                            field = STIL::artist;
                            break;
                        case 'c':
                        case 'C':
                            field = STIL::comment;
                            break;
                        default:
                            cout << "Wrong field. Assuming (A)ll." << endl;
                            field = STIL::all;
                            break;
                    }

                    cout << endl << "==== " << temp << ", Tune #" << tuneNo << " ====" << endl << endl;
                    cout << "---- GLOBAL  COMMENT ----" << endl;

                    tmpptr = myStil.getGlobalComment(temp);

                    if (tmpptr) {
                        cout << tmpptr;
                    }
                    else {
                        cout << "NONE!" << endl;
                    }

                    cout << "------ STIL  ENTRY ------" << endl;

                    tmpptr = myStil.getEntry(temp, tuneNo, field);

                    if (tmpptr) {
                        cout << tmpptr;
                    }
                    else {
                        cout << "NONE!" << endl;
                    }

                    cout << "---------- BUG ----------" << endl;

                    tmpptr = myStil.getBug(temp, tuneNo);

                    if (tmpptr) {
                        cout << tmpptr;
                    }
                    else {
                        cout << "NONE!" << endl;
                    }

                    cout << "==== END OF ENTRY ====" << endl << endl;
                }
            } while (*temp == '/');

            cout << "BYE!" << endl;
        }
    }

    return 0;
}
